tests/test_config.py

```diff
@@ -0,0 +1,131 @@
+        # directory. Without it, the ctime is the same, possibly because the
+        # deletion and creation coalesce somehow?
+        time.sleep(0.001)
+
+        function()
+
+        second_ctime = os.stat(getattr(config_, attribute)).st_ctime
+
+        self.assertNotEqual(first_ctime, second_ctime)
+
+    def test_singleton(self) -> None:
+        """Test that we get the same instance with subsequent calls."""
+        first_call_config = config.Config()
+        second_call_config = config.Config()
+
+        self.assertTrue(first_call_config is second_call_config)
+
+    def test_directory_creation(self) -> None:
+        """Test that we create the proper files mentioned in the config."""
+        config_ = config.Config()
+        self.assertTrue(os.path.exists(config_.data_dir))
+        self.assertTrue(os.path.exists(config_.cache_dir))
+        self.assertTrue(os.path.exists(config_.db_dir))
+
+    def test_all_attributes_initialised(self) -> None:
+        """Test that we create all the attributes in the default config."""
+        config_ = config.Config()
+        for key in config.DEFAULT_CONFIG.keys():
+            self.assertTrue(hasattr(config_, key))
+
+    def test_save_and_reload(self) -> None:
+        """
+        Test that reloading a config from file gives us the same config back, at least
+        with regards to the required attributes.
+        """
+        backup_config = config.Config()
+        backup_config.fullscreen = not backup_config.fullscreen
+        backup_config.save(".custom-config.json")
+
+        config.Config._instance = None
+
+        new_config = config.Config(".custom-config.json")
+
+        for key in config.DEFAULT_CONFIG:
+            self.assertTrue(getattr(backup_config, key) == getattr(new_config, key))
+
+    def test_malformed_config_load(self) -> None:
+        """Test that a malformed config is corrected on load."""
+        config_ = config.Config()
+        del config_.fullscreen
+        config_.save()
+
+        config.Config._instance = None
+        with self.assertLogs(level=logging.WARNING) as log_context_manager:
+            config.Config()
+        self.assertEqual(
+            log_context_manager.output,
+            [
+                "WARNING:root:Config is malformed, it does not have attribute "
+                "'fullscreen'. Using default value."
+            ]
+        )
+
+    def test_clear_cache(self) -> None:
+        """Test clearing cache does delete and recreate cache directory."""
+        self.assert_not_equal_ctime("cache_dir", config.clear_cache)
+
+    def test_clear_db(self) -> None:
+        """Test clearing database does delete and recreate database directory."""
+        self.assert_not_equal_ctime("db_dir", config.clear_db)
+
+    def test_clear_data(self) -> None:
+        """Test clearing data does delete and recreate data directory."""
+        self.assert_not_equal_ctime("data_dir", config.clear_data)
```

Comment:
I get what you're doing here, but I think instead of testing for deletion/recreation a better test might be to check whether these directories are empty. i.e. stick a file in, then assert that the directory is empty after the clear_data. 

The reasoning is that technically speaking, a clear_anything function ought to result in an empty directory. Whether that's done by deleting and recreating the directory, or by deleting every single file in a directory, is kinda irrelevant. The latter would result in the same desirable outcome but fail the test.

Reply:
That's what I meant to do at the start but the check for an empty directory doesn't actually work because `data_dir` is the parent to `cache_dir` and `db_dir` in the default config. That means they always get recreated before `clear_data` returns. I could do a recursive check for only empty directories in the tree starting at the directory of interest but that's a bit smelly to me.

So checking for empty makes sense in the case of db and cache but not data, and I figured since I needed a different check for data that I might as well use the same logic for the other two.

Reply:
I guess I could check for empty except for `cache_dir` and `db_dir` but this means that any time we add a default directory in the config, we need to check if it's inside another directory described in the config and if it is, we have to change the tests.

I'm not too happy with the ctime check either but I just didn't find any better option.

Reply:
I've taken a different approach now, using a sentinel file to ensure the directory was cleared.
Let me know if you think this is an appropriate alternative.